plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php

```diff
@@ -0,0 +1,217 @@
+
+	/**
+	 * Cache of classes already resolved.
+	 *
+	 * @var array
+	 */
+	protected array $resolved_cache;
+
+	/**
+	 * Initializes a new instance of the class.
+	 *
+	 * @param array $initial_resolved_cache Dictionary of class name => instance, to be used as the starting point for the resolved classes cache.
+	 */
+	public function __construct( array $initial_resolved_cache ) {
+		$this->resolved_cache = $initial_resolved_cache;
+	}
+
+	/**
+	 * Get an instance of a class.
+	 *
+	 * ContainerException will be thrown in these cases:
+	 *
+	 * - $class_name is outside the WooCommerce root namespace (and wasn't included in the initial resolve cache).
+	 * - The class referred by $class_name doesn't exist.
+	 * - Recursive resolution condition found.
+	 * - Reflection exception thrown when instantiating or initializing the class.
+	 *
+	 * A "recursive resolution condition" happens when class A depends on class B and at the same time class B depends on class A, directly or indirectly;
+	 * without proper handling this would lead to an infinite loop.
+	 *
+	 * Note that this method throwing ContainerException implies that code fixes are needed, it's not an error condition that's recoverable at runtime.
+	 *
+	 * @param string $class_name The class name.
+	 *
+	 * @return object The instance of the requested class.
+	 * @throws ContainerException Error when resolving the class to an object instance.
+	 * @throws \Exception Exception thrown in the constructor or in the 'init' method of one of the resolved classes.
+	 */
+	public function get( string $class_name ) {
+		$class_name    = trim( $class_name, '\\' );
+		$resolve_chain = array();
+		return $this->get_core( $class_name, $resolve_chain );
+	}
+
+	/**
+	 * Core function to get an instance of a class.
+	 *
+	 * @param string $class_name The class name.
+	 * @param array  $resolve_chain Classes already resolved in this resolution chain. Passed between recursive calls to the method in order to detect a recursive resolution condition.
+	 * @return object The resolved object.
+	 * @throws ContainerException Error when resolving the class to an object instance.
+	 */
+	protected function get_core( string $class_name, array &$resolve_chain ) {
+		if ( isset( $this->resolved_cache[ $class_name ] ) ) {
+			return $this->resolved_cache[ $class_name ];
+		}
+
+		// phpcs:disable WordPress.Security.EscapeOutput.ExceptionNotEscaped
+
+		if ( ! $this->is_class_allowed( $class_name ) ) {
+			throw new ContainerException( "Attempt to get an instance of class '$class_name', which is not in the " . self::WOOCOMMERCE_NAMESPACE . ' namespace. Did you forget to add a namespace import?' );
+		}
+
+		if ( ! class_exists( $class_name ) ) {
+			throw new ContainerException( "Attempt to get an instance of class '$class_name', which doesn't exist." );
+		}
+
+		// Account for the containers used by the Store API and Blocks.
+		if ( StringUtil::starts_with( $class_name, 'Automattic\WooCommerce\StoreApi\\' ) ) {
```

Comment:
There's [one instance where a core class depends on a Store API class](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/Orders/OrderAttributionBlocksController.php#L54). I guess it could be modified so that this particular dependency is explicitly grabbed from the Store API container whenever it's needed (and this might require adjusting some unit tests).

Anyway, the new container is supposed to be able retrieve any class from the `Automattic\WooCommerce` namespace, and it might be a bit weird if we had these two as exceptions (we would have to throw a "please use this other container" error); also the plan is to merge all three container in a single one as the next step of this project, so this explicit redirection to other containers is temporary.

Reply:
This is also a temporary issue during the transition, right?  Long-term do we plan to consolidate behind a single container, removing the `StoreApi` container?  Or are there specific methods/blockers that exist to doing that?

Reply:
In principle, the plan is indeed to remove the Store API and the Blocks container, and use this new simplified container for everything. This will probably mean that we'll need to add some additional functionality (e.g. right now the Blocks container is using instantiation callbacks) but the details (and blockers, if any) will surface after a more detailed look at the existing containers.